Include standard application layout and route helpers from main rails app.

winston opened the pull request:
Notes:
- removed engine's own layout
  - assumes that main app has bootstrap for styling to kick in (bad?)
- route helpers are included because layouts usually contains navigation links.

Also, the app that includes this engine would need to do this for all links in the layout,
where `main_app` is really a helper in Rails:

```
main_app.sign_in_path
```

---

Another way to do it:

```
module ActiveWaiter
  class ApplicationController < ::ApplicationController
    helper Rails.application.routes.url_helpers
  end
end
```

We can omit `layout` in this case, because then that's already assumed from inheriting from the main `ApplicationController`.

However, we still need the `helper`, and it doesn't solve the part about `main_app`.

~~I am considering this.. because if we just inherit from `ActionController::Base`, we lose things like `before_filter :authenticate_user!` on these waiter controllers.~~

Update: Not true anymore since the protection should be done within the parent app itself.

Example, we can use `devise`'s `authenticate` method in routes:

```
  authenticate :user do
    mount ActiveWaiter::Engine => "/active_waiter(/:id)"
  end
```

---

To do it such that we don't need to update the main parent app with `main_app.url`, we might need to remove the use of `mount` in `routes.rb` and write custom helpers to include the routes (like `devise`).

---

Thoughts?

winston committed:
Include standard application layout and route helpers from main rails app.

- route helpers are included because layouts usually contains navigation links.
- removed engine's own layout

choonkeat commented:
Since inheriting main app layout has more gotcha and require integration work, would prefer
1. keep the engine's plain `app/views/layouts/active_waiter/application.html.erb` file, and use it as default
2. add `ActiveWaiter.config` to activate [inheriting main app layout](https://github.com/choonkeat/active_waiter/blob/feature/inherit-parent-layout/app/controllers/active_waiter/application_controller.rb#L3-L4)
3. add documentation for `ActiveWaiter.config`
4. add compatibility changes (see below) for when `main_app` is not used in main layout (lazy, forgot, etc)

---

##### Now, if `main_app` is not done even if user chooses to activate inheriting main app layout...

Clicking on a link in the layout would hit something like
1. `/active_waiter/uid/users` instead of `/users` where the route is actually received by  `ActiveWaiter::JobsController#show`
2. OR `/active_waiter/uid/users/edit` instead of `/users/edit` where the route is a `404 Not Found` in the main app

If we change the setup instructions to setup `routes.rb`, then we can catch all routes for scenario (2) as well

``` ruby
# routes.rb
mount ActiveWaiter::Engine => "/active_waiter(/:id)(/*rewritepath)"
```

Now armed with `params[:rewritepath]` we can do a simple redirect to rectify the broken links in the layout

``` ruby
diff --git a/app/controllers/active_waiter/jobs_controller.rb b/app/controllers/active_waiter/jobs_controller.rb
index 6347b3b..60bed96 100644
--- a/app/controllers/active_waiter/jobs_controller.rb
+++ b/app/controllers/active_waiter/jobs_controller.rb
@@ -3,6 +3,11 @@ require_dependency "active_waiter/application_controller"
 module ActiveWaiter
   class JobsController < ApplicationController
     def show
+      if params[:rewritepath]
+        correct_request_uri = ["/#{params[:rewritepath]}", env['rack.request.query_string']].reject(&:blank?).join('?')
+        return redirect_to(correct_request_uri)
+      end
+
       data = ActiveWaiter.read(params[:id])
       return on_not_found(data) unless data.respond_to?(:[])
       return on_error(data)     if data[:error]
```

this would redirect `/active_waiter/uid/users/edit?abc=123&def=456` to `/users/edit?abc=123&def=456`

wdyt?

choonkeat commented:
> rewritepath

gotcha is, if the layout has links with different `method` like `delete` or `post`. sadly a redirect would not suffice for that.

winston commented:
> gotcha is, if the layout has links with different method like delete or post. sadly a redirect would not suffice for that.

Yea. The rewrite would be "difficult". This could be a solution instead? http://candland.net/2012/04/17/rails-routes-used-in-an-isolated-engine/

The others sound good. Will work on the `config`.

choonkeat commented:
> http://candland.net/2012/04/17/rails-routes-used-in-an-isolated-engine/

wow that's quite a big baggage to carry; i'd rather not honestly.

doing redirect on `rewritepath` is the farthest i'll go to fix the broken links "for free". otherwise, i'd rather not do anything.

winston committed:
Reinstate the default application view.

winston committed:
Configuration class for active_waiter.

winston committed:
Layout configured by configuration. Helper modules included if set.

winston committed:
Remove the ability to add helpers.

Initial plan was to have this configuration such that we can add `Rails.application.routes.url_helpers` to ActiveWaiter.
Adding that was only helpful in clearing errors when a parent app layout was used with routes originating from parent app.
However, the URL would still look like `host/active_waiter/.../whatever`, hence it didn't really solve all our problems.
To have the routes appear properly, you will still have to rename the routes used in the parent layout to `main_app.url`.

In light of that, this configuration is not very useful, so removing it for now.

winston committed:
Update readme with configuration bit.

winston commented:
- [x] keep the engine's plain app/views/layouts/active_waiter/application.html.erb file, and use it as default
- [x] add ActiveWaiter.config to activate inheriting main app layout
- [x] add documentation for ActiveWaiter.config
- ~~add compatibility changes (see below) for when main_app is not used in main layout (lazy, forgot, etc)~~

Done except the last point. Have added the note in the readme too. :eyes: :bow:

JuanitoFatas commented on README.md:
````diff
@@ -91,3 +90,18 @@ Optionally, you can also
+ 
+```ruby
+ActiveWaiter.configure do |config|
+  layout 'layouts/application
````
`config.layout = "app/views/layouts/application"`.

winston replied:
Updated.

winston committed:
Update readme.

choonkeat commented:
:+1:

JuanitoFatas commented:
- [x] This will need an Changelog entry @winston

winston committed:
Add change log.

winston commented:
Tests are failing only for Ruby 1.9.3 and Ruby 2.0 which we probably don't want to support, while passing for Ruby 2.1. Merging.

choonkeat commented:
1. Why are we abandoning a ruby version for test helper code?
2. Why are we using https://github.com/choonkeat/active_waiter/blob/3faa6af7c100bb3d6a9e5605809da04e7cb8a1a2/lib/active_waiter/configuration.rb and https://github.com/choonkeat/active_waiter/blob/3faa6af7c100bb3d6a9e5605809da04e7cb8a1a2/test/active_waiter/configuration_test.rb instead of #2 ?

choonkeat commented:
1. and test/controllers/active_waiter/jobs_controller_test.rb seem reverted from #2

winston commented:
:scream: I did a rebase from master.. Hmm..

winston commented:
> Why are we abandoning a ruby version for test helper code?

Yup. Let's fix that then. :bow:
